### PR TITLE
feat: use `best_platform` in a few places for cross-platform running

### PR DESCRIPTION
**Pull request by wolfv**
This is a draft. It adds a `best_platform` to most actions.

The idea is that we allow the selection of _another_ platform than the current one for "cross-platform" running. For example, Apple has implemented pretty good x86 emulation via Rosetta which means that most `osx-64` packages run very well on `osx-arm64`.

If a pixi.toml does not support `osx-arm64` (yet) we allow to fall back to `osx-64`.

This is especially important for `bioconda` as they are currently not building any packages for `osx-arm64`.

### CODE REVIEW

**Comment by wolfv** [comment]
fixes: https://github.com/prefix-dev/pixi/issues/875

cc @corneliusroemer would be interested in your thoughts.

**Comment by pavelzw** [comment]
How about a small warning when the platform doesn't macht your system platform?

**Comment by wolfv** [comment]
Yes, agree with the warning.

I had another thought - we could also do a setting as in `CONDA_SUBDIR=linux-64` where we would override the subdir per config / task / etc., e.g.

```
[platform."osx-arm64".options]
override_platform = "osx-64" # or platform_override?
```

But not sure if this is not too verbose.

Otherwise for configuration we would do somethign like:

```
platforms = ["osx-64", "osx-arm64:osx-64", "linux-*:linux-64"]
```

And map the platforms on each other like this.

**Comment by corneliusroemer** [comment]
I don't know enough about pixi to be of much help except for quite high-level feedback.

Fallback sounds good, warning too. Make sense to allow disabling fallback.

**Review by nichmor — COMMENTED** [review]
looks very good to me :)

File: src/cli/global/upgrade_all.rs [hunk 1]
```diff
@@ -30,6 +30,10 @@ pub struct Args {
 
     #[clap(flatten)]
     config: ConfigCli,
+
+    /// The platform to install the package for.
+    #[clap(long, default_value_t = Platform::current())]
+    platform: Platform,
```
Comment (ruben-arts) [inline src/cli/global/upgrade_all.rs:36]: On my machine this doesn't work if I use `emscripten-wasm32` 

Also wouldn't it make sense to by default get this from the environment itself? Just like the channels.

File: src/cli/global/install.rs [hunk 2]
```diff
@@ -44,6 +44,9 @@ pub struct Args {
     #[clap(short, long)]
     channel: Vec<String>,
 
+    #[clap(short, long, default_value_t = Platform::current())]
+    platform: Platform,
```
Comment (ruben-arts) [inline src/cli/global/install.rs:48]: The activation still uses the `Platform::current` is that how this should work?

**Review by ruben-arts — COMMENTED** [review]
This feature sounds logical but I would like some documentation and examples to try it out myself. As I can't get it to work from what I've been trying based on this pr.

File: src/cli/global/upgrade_all.rs (hunk 1 above)
Comment (ruben-arts) [reply src/cli/global/upgrade_all.rs:36]: @wolfv Could you create an wasm example to run some wasm stuff?

```toml
[project]
name = "wasm"
version = "0.1.0"
description = "Add a short description here"
authors = ["Wolf Vollprecht <w.vollprecht@gmail.com>"]
platforms = ["emscripten-wasm32", "osx-arm64"]
channels = ["conda-forge"]

[feature.host]
channels = ["conda-forge"]
platforms = ["osx-arm64"]
dependencies = { python = "*" }

[feature.wasm]
channels = ["https://repo.mamba.pm/emscripten-forge"]
dependencies = { python = "*", pillow = "*", numpy = "*" }
platforms = ["emscripten-wasm32"]

[environments]
wasm = ["wasm"]
host = ["host"]
```

Comment (wolfv) [reply src/cli/global/upgrade_all.rs:36]: Let's do it after the merge so that Torsten can help :)

**Review by ruben-arts — APPROVED** [review]

**Pull request merged**